### PR TITLE
time unit not match in ThreadMetricUtil.cpp

### LINKED ISSUE #5595: time unit not match in ThreadMetricUtil.cpp
## Bug Report

Please answer these questions before submitting your issue. Thanks!

https://github.com/pingcap/tiflash/blob/f094fa5b27156789b3ba629360396d887939de78/dbms/src/Common/ThreadMetricUtil.cpp#L27-L28

max_thds_metric_reset_interval is in seconds.

### PR DESCRIPTION
**Pull request by Lloyd-Pottiger**
Signed-off-by: Lloyd-Pottiger <yan1579196623@gmail.com>

### What problem does this PR solve?

Issue Number: close #5595

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Review by bestwoody — APPROVED** [review]

**Review by ywqzzy — APPROVED** [review]
LGTM

File: dbms/src/Common/ThreadMetricUtil.cpp [hunk 1]
```diff
@@ -25,7 +25,7 @@ namespace DB
 bool tryToResetMaxThreadsMetrics()
 {
     UInt64 now_ts = clock_gettime_ns_adjusted(last_max_thds_metric_reset_ts, CLOCK_MONOTONIC);
-    if (now_ts > last_max_thds_metric_reset_ts + max_thds_metric_reset_interval)
+    if (now_ts > last_max_thds_metric_reset_ts + max_thds_metric_reset_interval * 1000000000ULL)
```
Comment (fuzhe1989) [inline dbms/src/Common/ThreadMetricUtil.cpp:28]: how about all use `std::chrono::duration` for comparing (so there won't be any silent unit mismatch)?

Comment (Lloyd-Pottiger) [reply dbms/src/Common/ThreadMetricUtil.cpp:28]: yep, cool. `last_max_thds_metric_reset_ts` need to be `atomic` here, using `std::chrono::duration` will increase the overhead?

Comment (fuzhe1989) [reply dbms/src/Common/ThreadMetricUtil.cpp:28]: well, `last_max_thds_metric_reset_ts` is fetched twice here so the two fetched values could be different, will it be an issue?

my suggestion is: fetch `last_max_thds_metric_reset_ts` before `clock_gettime_ns_adjusted` and store the value as a `std::chrono::duration`.

Comment (Lloyd-Pottiger) [reply dbms/src/Common/ThreadMetricUtil.cpp:28]: lgtm

Comment (Lloyd-Pottiger) [reply dbms/src/Common/ThreadMetricUtil.cpp]: @fuzhe1989 done, PTAL again

File: dbms/src/Common/ThreadMetricUtil.cpp [hunk 2]
```diff
@@ -18,16 +18,18 @@
 #include <common/types.h>
 
 std::atomic<UInt64> last_max_thds_metric_reset_ts{0};
-const UInt64 max_thds_metric_reset_interval = 60; //60s
+const std::chrono::duration<UInt64> max_thds_metric_reset_interval{60}; // 60s
```
Comment (bestwoody) [inline dbms/src/Common/ThreadMetricUtil.cpp:21]: is `second` the default unit of duration ?

File: dbms/src/Common/ThreadMetricUtil.cpp [hunk 3]
```diff
@@ -18,16 +18,18 @@
...
 
 namespace DB
 {
 bool tryToResetMaxThreadsMetrics()
 {
-    UInt64 now_ts = clock_gettime_ns_adjusted(last_max_thds_metric_reset_ts, CLOCK_MONOTONIC);
-    if (now_ts > last_max_thds_metric_reset_ts + max_thds_metric_reset_interval)
+    using microseconds = std::chrono::duration<UInt64, std::micro>;
+    UInt64 last_max_thds_metric_reset_ts_tmp = last_max_thds_metric_reset_ts.load(std::memory_order_relaxed);
+    UInt64 now_ts = clock_gettime_ns_adjusted(last_max_thds_metric_reset_ts_tmp, CLOCK_MONOTONIC);
```
Comment (bestwoody) [inline dbms/src/Common/ThreadMetricUtil.cpp:29]: the unit of returned val is `ns`, however , we use `microseconds`

Comment (Lloyd-Pottiger) [reply dbms/src/Common/ThreadMetricUtil.cpp:29]: sorry, my bad

File: dbms/src/Common/ThreadMetricUtil.cpp [hunk 4]
```diff
@@ -16,18 +16,21 @@
 #include <Common/ThreadMetricUtil.h>
 #include <Common/TiFlashMetrics.h>
 #include <common/types.h>
+#include <ratio>
```
Comment (Lloyd-Pottiger) [inline dbms/src/Common/ThreadMetricUtil.cpp:19]:
```suggestion
```

File: dbms/src/Common/ThreadMetricUtil.cpp (hunk 3 above)
Comment (Lloyd-Pottiger) [reply dbms/src/Common/ThreadMetricUtil.cpp:29]: fixed

**Review by bestwoody — APPROVED** [review]

File: dbms/src/Common/ThreadMetricUtil.cpp [hunk 5]
```diff
@@ -18,16 +18,18 @@
...
-const UInt64 max_thds_metric_reset_interval = 60; //60s
+const std::chrono::duration<UInt64> max_thds_metric_reset_interval{60}; // 60s
 
 namespace DB
 {
 bool tryToResetMaxThreadsMetrics()
 {
-    UInt64 now_ts = clock_gettime_ns_adjusted(last_max_thds_metric_reset_ts, CLOCK_MONOTONIC);
-    if (now_ts > last_max_thds_metric_reset_ts + max_thds_metric_reset_interval)
+    using nanoseconds = std::chrono::duration<UInt64, std::nano>;
```
Comment (fuzhe1989) [inline dbms/src/Common/ThreadMetricUtil.cpp:27]:
```suggestion
```

File: dbms/src/Common/ThreadMetricUtil.cpp [hunk 6]
```diff
@@ -18,16 +18,18 @@
 #include <common/types.h>
 
 std::atomic<UInt64> last_max_thds_metric_reset_ts{0};
```
Comment (fuzhe1989) [inline dbms/src/Common/ThreadMetricUtil.cpp:20]:
```suggestion
using Clock = std::chrono::steady_clock;
using TimePoint = Clock::time_point;

std::atomic<TimePoint> last_max_thds_metric_reset_ts{};
```

File: dbms/src/Common/ThreadMetricUtil.cpp [hunk 7]
```diff
@@ -18,16 +18,18 @@
...
+const std::chrono::duration<UInt64> max_thds_metric_reset_interval{60}; // 60s
 
 namespace DB
 {
 bool tryToResetMaxThreadsMetrics()
 {
-    UInt64 now_ts = clock_gettime_ns_adjusted(last_max_thds_metric_reset_ts, CLOCK_MONOTONIC);
-    if (now_ts > last_max_thds_metric_reset_ts + max_thds_metric_reset_interval)
+    using nanoseconds = std::chrono::duration<UInt64, std::nano>;
+    UInt64 last_max_thds_metric_reset_ts_tmp = last_max_thds_metric_reset_ts.load(std::memory_order_relaxed);
```
Comment (fuzhe1989) [inline dbms/src/Common/ThreadMetricUtil.cpp:28]:
```suggestion
    auto last_max_thds_metric_reset_ts_tmp = last_max_thds_metric_reset_ts.load(std::memory_order_relaxed);
```

File: dbms/src/Common/ThreadMetricUtil.cpp [hunk 8]
```diff
@@ -18,16 +18,18 @@
...
+    UInt64 now_ts = clock_gettime_ns_adjusted(last_max_thds_metric_reset_ts_tmp, CLOCK_MONOTONIC);
```
Comment (fuzhe1989) [inline dbms/src/Common/ThreadMetricUtil.cpp:29]:
```suggestion
    auto now = std::max(Clock::now(), last_max_thds_metric_reset_ts_tmp );
```

File: dbms/src/Common/ThreadMetricUtil.cpp [hunk 9]
```diff
@@ -18,16 +18,18 @@
...
+    if (nanoseconds(now_ts) > nanoseconds(last_max_thds_metric_reset_ts_tmp) + max_thds_metric_reset_interval)
```
Comment (fuzhe1989) [inline dbms/src/Common/ThreadMetricUtil.cpp:30]:
```suggestion
    if (now > last_max_thds_metric_reset_ts_tmp + max_thds_metric_reset_interval)
```

File: dbms/src/Common/ThreadMetricUtil.cpp [hunk 10]
```diff
@@ -18,16 +18,18 @@
...
     {
-        last_max_thds_metric_reset_ts = now_ts;
+        last_max_thds_metric_reset_ts.store(now_ts, std::memory_order_relaxed);
```
Comment (fuzhe1989) [inline dbms/src/Common/ThreadMetricUtil.cpp:32]:
```suggestion
        last_max_thds_metric_reset_ts.store(now, std::memory_order_relaxed);
```

**Review by fuzhe1989 — COMMENTED** [review]
Sorry for late response. IMO the merged version doesn't provide enough type safety and should be refined.

**Pull request merged**